src/jit/morph.cpp

```diff
@@ -1248,158 +1248,110 @@ void fgArgInfo::RemorphReset()
-            argx     = curArgTabEntry->parent->Current();
-            isRegArg = (argx->gtFlags & GTF_LATE_ARG) != 0;
-        }
-        else
-        {
-            argx     = curArgTabEntry->node;
-            isRegArg = true;
-        }
-
-        if (curArgTabEntry->argNum == argNum)
-        {
-            break;
-        }
-
-        if (isRegArg)
-        {
-            regArgInx++;
-        }
-    }
+    bool isLateArg = (curArgTabEntry->lateArgInx != (unsigned)-1);
 
-    nextSlotNum = roundUp(nextSlotNum, alignment);
-
-    assert(curArgTabEntry->argNum == argNum);
+    noway_assert(curArgTabEntry->parent != nullptr);
+    assert(curArgTabEntry->parent->OperIsList());
+    assert(curArgTabEntry->parent->Current() == node);
+    nextSlotNum = (unsigned)roundUp(nextSlotNum, curArgTabEntry->alignment);
     assert(curArgTabEntry->slotNum == nextSlotNum);
-    assert(curArgTabEntry->numSlots == numSlots);
-    assert(curArgTabEntry->alignment == alignment);
-    assert(curArgTabEntry->parent == parent);
-    assert(parent->OperIsList());
+    assert((isLateArg && ((node->gtFlags & GTF_LATE_ARG) != 0)) ||
+           (!isLateArg && ((node->gtFlags & GTF_LATE_ARG) == 0)));
 
-#if FEATURE_FIXED_OUT_ARGS
     if (curArgTabEntry->node != node)
     {
-        if (isRegArg)
+#if FEATURE_FIXED_OUT_ARGS
+        if (isLateArg)
         {
-            GenTree* argx     = nullptr;
-            unsigned regIndex = 0;
+            GenTree* argx = nullptr;
+
+            // First, determine the lateArgInx, if it has not been set in curArgTabEntry.
+            unsigned lateArgInx = curArgTabEntry->lateArgInx;
+            if (lateArgInx == -1)
+            {
+                for (unsigned i = 0; i < this->argCount; i++)
+                {
+                    if (curArgTabEntry == argTable[i])
+                    {
+                        lateArgInx = i;
+                        break;
+                    }
+                }
+            }
+            assert(lateArgInx != -1);
 
-            /* process the register argument list */
-            for (GenTreeArgList *list = callTree->gtCall.gtCallLateArgs; list; list = list->Rest(), regIndex++)
+            // Now, traverse the late argument list to find this argument so that we can update it.
+            unsigned listInx = 0;
+            for (GenTreeArgList *list = callTree->gtCall.gtCallLateArgs; list; list = list->Rest(), listInx++)
             {
                 argx = list->Current();
                 assert(!argx->IsArgPlaceHolderNode()); // No place holders nodes are in gtCallLateArgs;
-                if (regIndex == regArgInx)
+                if (listInx == lateArgInx)
                 {
                     break;
                 }
             }
-            assert(regIndex == regArgInx);
-            assert(regArgInx == curArgTabEntry->lateArgInx);
+            assert(listInx == lateArgInx);
+            assert(lateArgInx == curArgTabEntry->lateArgInx);
```

Comment:
Is lateArgInx assigned into curArgTabEntry if it is -1 originally? Seems like this assert may compare -1 and lateArgInx

Reply:
I did some simplification in this method, and it looks like more is needed - that is, I think that the `lateArgInx != -1` should always be true if `isLateArg`. I'll take a deeper look at this.